### ISSUE TITLE
Dynamic Snippets

### ISSUE THREAD

**Issue by rgardner** [atom/snippets#33]
I'd like to dynamically include the current date/time in a snippet for markdown.
The snippet would look something like this.

``` cson
'.source.gfm':
  'Header':
    'prefix': 'header'
    'body': 'My name - `Date.today.strftime('%D')`'
```

which would print:

My name - 4/2/14
Date.today.strftime('%D') is Ruby, the equivalent js a little longer (unless there's an easier way):

``` js
var d = new Date();
var curr_date = d.getDate();
var curr_month = d.getMonth();
curr_month++;
var curr_year = d.getFullYear();
console.log(curr_month + "/" + curr_date + "/" + curr_year);
```

Thoughts?
I originally posted this [on the atom forum](http://discuss.atom.io/t/dynamic-snippet-text-in-snippets-cson/6920)

**Comment by pathawks** [comment]
Related: atom/snippets#83

**Comment by joefitzgerald** [comment]
#106 + an autocomplete-plus provider could address this, once I am done with #106 and https://github.com/atom-community/autocomplete-plus/pull/245.

**Comment by jb** [comment]
So this still hasn't been implemented? What needs doing? Someone needs to fix pull request #83?

<hr>

@rgardner: I would use [Moment.js](http://momentjs.com/):

    require moment
    moment().format("YYYY-MM-DD")

<small>Please don't format dates jumbled up. You wouldn't write `50:43:08`, so please don't write `05-21-98`. It's annoying to anyone who either **a)** wants to sort the dates programmatically or **b)** is not American. I.e: most of the world.</small>

**Comment by joefitzgerald** [comment]
This is done and is not dependent on #83. It was implemented in https://github.com/atom/snippets/pull/110. [`autocomplete-plus`](https://github.com/atom/autocomplete-plus) uses this API to allow autocomplete providers to emit dynamic snippets for their suggestions. You can do so also, outside the context of an autocomplete-plus provider.

**Comment by joefitzgerald** [comment]
I suppose me stating this is done isn't fully accurate. The ability for packages to integrate with snippets to add dynamic snippets exists. Autocomplete-plus allows you to do so based on the scopechain underneath the current cursor. In either case, it requires an additional package to drive the requirements you're describing above, but the path is paved for that to happen.

#83 would allow people to add their own snippets and drive dynamic behavior without any other packages. You could also look to add this to [`autocomplete-snippets`](https://github.com/atom/autocomplete-snippets).

**Comment by imcoddy** [comment]
+1
Really wish to see this feature coming out as it is a bit annoying to add a timestamp to a file by hand

**Comment by twz915** [comment]
+1 really wish to have this function to add a timestamp too

**Comment by Strae** [comment]
+1

**Comment by pathawks** [comment]
@Strae Rather than just adding a `+1` style comment, Github now allows you to add an emoji reaction to an issue. Giving an issue a :+1: is more appropriate than leaving the equivalent comment.

**Comment by calebmeyer** [comment]
Alternative way to accomplish this that's mildly more flexible:
```cson
'.source.gfm':
  'Header':
    'prefix': 'header'
    'body': 'My name - $date'
    'variables':
      'date': '''
        require moment
        moment().format("YYYY-MM-DD")
      '''
```

If `$date` seems weird then we could put date in backticks instead

I say this way is more flexible because it allows you to reuse the variables more than once.

**Comment by Strae** [comment]
@calebmeyer this is great, requiring modules also let us keep the snippet file clean

**Comment by calebmeyer** [comment]
@Strae yeah, that would be even better than what I had there, although we'd have to be clear about when things were evaluated. Ideally every variable is evaluated lazily when it's needed; that way you don't end up with yesterday's date if you leave atom open overnight.

Scoping rules would be pretty easy, too. Just search up the tree until you find a variables block that contains the variable you're looking for.

**Comment by Strae** [comment]
You're right.
In my case for example, I just want to add a random string to a console.log() invoke, so I can easily search the code for that string when I forget to remove the quick'n'dirty logs.. this implies that the code that generate the string must be evalutated everytime the snippet is used.

**Comment by Vanav** [comment]
I'm sorry, is there any workaround to insert current date and time as a part of snippet? Using Service API?

**Comment by chrishiestand** [comment]
@Vanav I've used the `date` package to insert the current date/time. But not as part of a larger snippet.

**Comment by Redsandro** [comment]
I'd love to see something like this again. The settings used to be `eval()`'ed, making this possible, but it slowed down starting Atom.

So it was removed in 2014: https://github.com/atom/atom/pull/4971

**Comment by rhoglund** [comment]
+1

**Comment by metalelf0** [comment]
Sorry it this has already been answered anywhere else, but I still haven't found it. Right now is there any way to insert current date / time as part of an atom snippet? If so, could anybody post an example snippet? Thanks in advance!

**Comment by joefitzgerald** [comment]
Yes. You can do this in an autocomplete-plus provider (see generateSnippet in https://github.com/joefitzgerald/go-plus/blob/master/lib/autocomplete/gocodeprovider.js) or directly by copying the logic from replaceTextWithMatch in https://github.com/atom/autocomplete-plus/blob/master/lib/autocomplete-manager.js.

**Comment by akshitarora** [comment]
:+1:
